### ISSUE TITLE
Editor: Address accessibility concerns

### ISSUE THREAD

**Issue by alisterscott** [Automattic/wp-calypso#697]
Originally raised by @codebykat 

Comment from @davidakennedy:

I'm using this [set of tests as my broad guideline](http://davidakennedy.com/2014/10/31/web-accessibility-in-60-seconds/). It's not everything, but it's the basics and some of the more foundational items when it comes to accessibility. I may dip into more advanced items if it's warranted. But I figure we can came back to more advanced items at a later time.

**Browser**: Chrome Version 46.0.2490.80 (64-bit) and Safari Version 9.0.1 (11601.2.7.2)
**Screen Reader**: Voiceover
**Tools used**: my keyboard, Wave Toolbar and Google Accessibility Tools for Chrome.

## Overall

Nice work! There are some issues, but most of of them are easy wins. I was able to navigate the TinyMCE editor part of the editor pretty well with a screen reader and keyboard.

## Structure

Some of these items likely extend beyond the Editor itself, but I'll mention them anyway.

* [ ] A skip link to the main content area would be nice. That makes it easier for a keyboard user to get to the meat of the page.
* [ ] Making the `<div id="content"` div a `<main>` element would add an additional navigational ARIA landmark for assistive technology users, and be valuable.
* [ ] There are no headings on this page at all. That's not bad, per se. Sighted users pick up a lot of visual cues and hierarchy because of the size of type, shades of color and certain shapes. It may be worth adding some visually hidden headings: An `h1` that says Add new post, an one `h2`, above the sidebar area that says Edit post details, or something like that. This way, screen reader users have another shortcut way of navigating the page – by heading elements. It also gives them hierarchy, like sighted user get.

## Basic Testing

Missing Labels:

All form controls need a `label` element. Associating the control with the label helps assistive technology users to know what to do with the control. It's also invalid HTML if labels are missing. See this article for any [additional help](http://webaim.org/techniques/forms/controls).

* [x] post slug input in Sharing section of sidebar.
* [x] except input in More Option section of sidebar.
* [x] delete post button has no text inside button. Either add and hide off screen or use `aria-label` to add accessible name.
* [x] post title input in main editor area.
* [x] All of the buttons in the web preview modal need text of some sort. Either add and hide off screen or use `aria-label` to add accessible name.
* [ ] The web preview iframe needs a title attribute.

## Non-Semantic elements

* [ ] edit permalink button in main editor area should be a `button` because keyboard users cannot access `div` elements. They aren't keyboard accessible by default.

## Keyboard Testing

* [ ] Better visible `:focus` styles needed around Drafts link in sidebar.
* [ ] Better visible `:focus` styles on Show advanced status details button in sidebar.
* [ ] I have similar [confusion as this user](https://horizonfeedback.wordpress.com/2015/09/14/call-for-testing-editor/comment-page-1/#comment-144) when using the keyboard alone: 
* [ ] Accordions in sidebar cannot be accessed via a keyboard. Also, the source order seems off.

## Screen Reader Testing

Most of these also apply to keyboard testing as well.

Two of the bigger issues I see on this page:

* [ ] some content that is visually hidden isn't hidden for screen readers. That means they can access items that they shouldn't be able to at certain times. These include the accordions in the sidebar and the web preview. Those items should be `display: none;` until they are brought onscreen by user action. Otherwise, it's very confusing.
* [ ] When the web preview modal is brought into view, focus should be programmatically shifted to it. When it's closed, focus should shift back to the preview button. This way, a keyboard or screen reader user can use this feature and not lose their place. See one example of a [dialog modal](http://heydonworks.com/practical_aria_examples/#warning-dialog).

## Other Misc Advanced Items

These are just a few issues that get into more advanced territory, but that I wanted to mention since I noticed them.

* [ ] The links for the Visual and HTML tabs of the Editor that have `role=radio` need a `aria-checked="false"` or `aria-checked="true"` depending on their state. That may be a Core issue with TinyMCE.
* [ ] Disabling zoom with `<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">` is considered bad practice for accessibility. Mainly because many times, users with poor vision need to pinch and zoom to see items and read things.
* [ ] Auto-saving is not indicated to a visually impaired person, and  after auto-save there is no longer a "save" button. So the author only gets a "publish" button, and they probably are unaware it has already been saved. @rralian brought this up. We could solve this by using `aria-live` to communicate to assistive technology when a post/page is saved. See this [example](http://heydonworks.com/practical_aria_examples/#offline-alert).
* [ ] When already on the Editor page, clicking "New Post" and selecting a different site from the site-picker via keyboard controls will update the url, but there is no feedback and the site-picker remains open. @rralian brought this up too.

**Comment by amandarush** [comment]
Should we maybe make all of these separate issues, so that devs can fix
them one-by-one and check them off?





*From:* Alister Scott [mailto:notifications@github.com]
*Sent:* Tuesday, November 24, 2015 5:38 PM
*To:* Automattic/wp-calypso
*Subject:* [wp-calypso] Editor: Address accessibility concerns (#697)



Originally raised by @codebykat <https://github.com/codebykat>

Comment from @davidakennedy <https://github.com/davidakennedy>:

I'm using this set of tests as my broad guideline
<http://davidakennedy.com/2014/10/31/web-accessibility-in-60-seconds/>.
It's not everything, but it's the basics and some of the more foundational
items when it comes to accessibility. I may dip into more advanced items if
it's warranted. But I figure we can came back to more advanced items at a
later time.

*Browser*: Chrome Version 46.0.2490.80 (64-bit) and Safari Version 9.0.1
(11601.2.7.2)
*Screen Reader*: Voiceover
*Tools used*: my keyboard, Wave Toolbar and Google Accessibility Tools for
Chrome.
Overall

Nice work! There are some issues, but most of of them are easy wins. I was
able to navigate the TinyMCE editor part of the editor pretty well with a
screen reader and keyboard.
Structure

Some of these items likely extend beyond the Editor itself, but I'll
mention them anyway.

   - [ ] A skip link to the main content area would be nice. That makes it
   easier for a keyboard user to get to the meat of the page.
   - [ ] Making the <div id="content" div a <main> element would add an
   additional navigational ARIA landmark for assistive technology users, and
   be valuable.
   - [ ] There are no headings on this page at all. That's not bad, per se.
   Sighted users pick up a lot of visual cues and hierarchy because of the
   size of type, shades of color and certain shapes. It may be worth adding
   some visually hidden headings: An h1 that says Add new post, an one h2,
   above the sidebar area that says Edit post details, or something like that.
   This way, screen reader users have another shortcut way of navigating the
   page – by heading elements. It also gives them hierarchy, like sighted user
   get.

Basic Testing

Missing Labels:

All form controls need a label element. Associating the control with the
label helps assistive technology users to know what to do with the control.
It's also invalid HTML if labels are missing. See this article for any
additional
help <http://webaim.org/techniques/forms/controls>.

   - [X] post slug input in Sharing section of sidebar.
   - [X] except input in More Option section of sidebar.
   - [X] delete post button has no text inside button. Either add and hide
   off screen or use aria-label to add accessible name.
   - [X] post title input in main editor area.
   - [X] All of the buttons in the web preview modal need text of some
   sort. Either add and hide off screen or use aria-label to add accessible
   name.
   - [ ] The web preview iframe needs a title attribute.

Non-Semantic elements

   - [ ] edit permalink button in main editor area should be a button
   because keyboard users cannot access div elements. They aren't keyboard
   accessible by default.

Keyboard Testing

   - [ ] Better visible :focus styles needed around Drafts link in sidebar.
   - [ ] Better visible :focus styles on Show advanced status details
   button in sidebar.
   - [ ] I have similar confusion as this user
   <https://horizonfeedback.wordpress.com/2015/09/14/call-for-testing-editor/comment-page-1/#comment-144>
   when using the keyboard alone:
   - [ ] Accordions in sidebar cannot be accessed via a keyboard. Also, the
   source order seems off.

Screen Reader Testing

Most of these also apply to keyboard testing as well.

Two of the bigger issues I see on this page:

   - [ ] some content that is visually hidden isn't hidden for screen
   readers. That means they can access items that they shouldn't be able to at
   certain times. These include the accordions in the sidebar and the web
   preview. Those items should be display: none; until they are brought
   onscreen by user action. Otherwise, it's very confusing.
   - [ ] When the web preview modal is brought into view, focus should be
   programmatically shifted to it. When it's closed, focus should shift back
   to the preview button. This way, a keyboard or screen reader user can use
   this feature and not lose their place. See one example of a dialog modal
   <http://heydonworks.com/practical_aria_examples/#warning-dialog>.

Other Misc Advanced Items

These are just a few issues that get into more advanced territory, but that
I wanted to mention since I noticed them.

   - [ ] The links for the Visual and HTML tabs of the Editor that have
   role=radio need a aria-checked="false" or aria-checked="true" depending
   on their state. That may be a Core issue with TinyMCE.
   - [ ] Disabling zoom with <meta name="viewport"
   content="width=device-width, initial-scale=1, maximum-scale=1"> is
   considered bad practice for accessibility. Mainly because many times, users
   with poor vision need to pinch and zoom to see items and read things.
   - [ ] Auto-saving is not indicated to a visually impaired person, and
   after auto-save there is no longer a "save" button. So the author only gets
   a "publish" button, and they probably are unaware it has already been
   saved. @rralian <https://github.com/rralian> brought this up. We could
   solve this by using aria-live to communicate to assistive technology
   when a post/page is saved. See this example
   <http://heydonworks.com/practical_aria_examples/#offline-alert>.
   - [ ] When already on the Editor page, clicking "New Post" and selecting
   a different site from the site-picker via keyboard controls will update the
   url, but there is no feedback and the site-picker remains open. @rralian
   <https://github.com/rralian> brought this up too.

—
Reply to this email directly or view it on GitHub
<https://github.com/Automattic/wp-calypso/issues/697>.

**Comment by davidakennedy** [comment]
@amandarush We talked about that a bit, and it didn't seem to matter to the developers. But it probably makes sense, particularly to break up the tasks that are more general from the editor-related ones.

**Comment by alisterscott** [comment]
@amandarush @davidakennedy I have now split this into individual issues for each of the items to more easily track/manage/assign/fix. I will close this master issue, there are links to the new issues above.
